x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/query/SparseTermsQueryBuilder.java

```diff
@@ -0,0 +1,160 @@
+    private static final ParseField SCORES = new ParseField("scores");
+    private static final ParseField TERMS = new ParseField("terms");
+
+    @SuppressWarnings("unchecked")
+    public static final ConstructingObjectParser<SparseTermsQueryBuilder, Void> PARSER = new ConstructingObjectParser<>(
+        "sparse_terms",
+        false,
+        a -> new SparseTermsQueryBuilder((String) a[0], (List<String>) a[1], (List<Float>) a[2])
+    );
+
+    static {
+        PARSER.declareString(ConstructingObjectParser.constructorArg(), FIELD);
+        PARSER.declareStringArray(ConstructingObjectParser.constructorArg(), TERMS);
+        PARSER.declareFloatArray(ConstructingObjectParser.constructorArg(), SCORES);
+        AbstractQueryBuilder.declareStandardFields(PARSER);
+    }
+
+    private final List<String> terms;
+    private final List<Float> scores;
+    private final String fieldName;
+
+    public SparseTermsQueryBuilder(String fieldName, List<String> terms, List<Float> scores) {
+        this.fieldName = requireNonNull(fieldName, FIELD);
+        this.terms = requireNonNull(terms, TERMS);
+        this.scores = requireNonNull(scores, SCORES);
+        if (this.scores.size() != this.terms.size()) {
+            throw new IllegalArgumentException(format("[%s] length must equal [%s]", TERMS.getPreferredName(), SCORES.getPreferredName()));
+        }
+    }
+
+    public SparseTermsQueryBuilder(StreamInput in) throws IOException {
+        super(in);
+        this.terms = in.readStringList();
+        this.scores = in.readList(StreamInput::readFloat);
+        this.fieldName = in.readString();
+    }
+
+    public List<String> getTerms() {
+        return terms;
+    }
+
+    public List<Float> getScores() {
+        return scores;
+    }
+
+    @Override
+    public String getWriteableName() {
+        return NAME;
+    }
+
+    @Override
+    public Version getMinimalSupportedVersion() {
+        return Version.CURRENT;
+    }
+
+    @Override
+    protected void doWriteTo(StreamOutput out) throws IOException {
+        out.writeStringCollection(terms);
+        out.writeCollection(scores, StreamOutput::writeFloat);
+        out.writeString(fieldName);
+    }
+
+    @Override
+    protected void doXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.startObject(getName());
+        builder.field(FIELD.getPreferredName(), fieldName);
+        builder.field(SCORES.getPreferredName(), scores);
+        builder.field(TERMS.getPreferredName(), terms);
+        boostAndQueryNameToXContent(builder);
+        builder.endObject();
+    }
+
+    @Override
+    protected Query doToQuery(SearchExecutionContext context) {
+        MappedFieldType fieldType = context.getFieldType(fieldName);
+        if (fieldType == null) {
+            throw new IllegalStateException("Rewrite first");
+        }
+        if (fieldType instanceof RankFeaturesFieldMapper.RankFeaturesFieldType == false) {
```

Comment:
Another option here is `fieldType.name().equals("rank_features")` that way we don't have to worry about importing the module or exporting the class.